src/env-replace.js

```diff
@@ -0,0 +1,138 @@
+module.exports = envReplace
+
+/**
+ * This will attempt to resolve the value of any env variables that are inside
+ * this string. For example, it will transform this:
+ * cross-env FOO=$NODE_ENV BAR=\\$NODE_ENV echo $FOO ${BAR}
+ * Into this:
+ * FOO=development BAR=$NODE_ENV echo $FOO
+ * (Or whatever value the variable NODE_ENV has)
+ * Note that this function is only called with the right-side portion of the
+ * env var assignment, so in that example, this function would transform
+ * the string "$NODE_ENV" into "development"
+ *
+ * To specify a default value for a variable, use the ${ENV_VAR_NAME:-default value} syntax.
+ *
+ * The default value syntax can be nested to resolve multiple environment variables:
+ *   ${VAR1:-${VAR2:-${VAR3:-default if none set}}}
+ *
+ * @param {String} value The command/value to replace
+ * @param {Object} env optional environment object (defaults to process.env)
+ * @param {boolean} winEnvReplace If true, replace environment variables with '%NAME%' instead of their actual value
+ * @returns {String} the value with replacements
+ **/
+// state machine requires slightly more complexity than normally required
+// eslint-disable-next-line complexity
+function envReplace(value, env = process.env, winEnvReplace = false) {
```

Comment:
This is amazing, but I *really* don't want to maintain this function. And environment variable expansion syntax is not something I'm very familiar with.

Would you agree to help maintain this code if I merge?

Reply:
An alternative might be to use a PEG parser generator or something. Let me see if I can come up with something, just to see what it looks like. Total props for writing this though @yinzara, not having a dig at all. I've written my fare share of this kind of thing, and they are hard!

Reply:
Honestly I'm happy to help maintain what I've written. Love to see your PEG variation if you would like. Up you you

Reply:
Sure, got something mostly working now, but need to support escape characters and stuff.

Reply:
This is what I've got in mind. I'm not totally done yet and I've got to head out now for a bit, but this is pretty close. Just need to sort out a few final cases like `${BOB:-hello "    there"}` and make sure it behaves like bash. You can load this up and play with it here: https://pegjs.org/online.

```
{
  const process = {
    env: {
      BOB: "Hello!"
    }
  };
  
  function getEnv(name, fallback) {
     const env = process.env[name];
     return (env == null || env === '')
     	? (fallback == null ? '' : fallback)
        : env;
  }
}

start
  = String

String = QuotedString / UnquotedString

EnvVar
  = "${" _ name:VarName _ (":-" / ":=") fallback:FallbackValue "}" { return getEnv(name, fallback); }
  / "${" _ name:VarName _ "}" { return getEnv(name); }
  / "$" name:VarName { return getEnv(name); }
  / "$" { return "$"; }

FallbackValue
  = QuotedString
  / val:((BracketStringChar / EnvVar)*) { return val.join('').trim(); }

VarName "environment variable name"
  = [a-zA-Z_][a-zA-z_0-9]+ { return text(); }

_ "whitespace"
  = [ \t\n\r\v]*

QuotedString
  = '"' val:(DoubleStringChar / EnvVar)* '"' { return val.join(''); }
  / "'" val:(SingleStringChar / EnvVar)* "'" { return val.join(''); }
  
UnquotedString
  = val:(BareStringChar / EnvVar)* { return val.join(''); }
  
DoubleStringChar
  = '\\' escaped:("$" / '"') { return escaped; }
  / !("$" / '"') char:. { return char; }

SingleStringChar
  = "\\" escaped:("$" / "'") { return escaped; }
  / !("$" / "'") char:. { return char; }

BracketStringChar
  = "\\" escaped:("$" / "}") { return escaped; }
  / !("$" / "}") char:. { return char; }

BareStringChar
  = "\\" escaped:("$") { return escaped; }
  / !("$") char:. { return char; }
```

Reply:
Is this something you'd like me to continue exploring? Personally I find these grammars are a bit easier to fiddle with and maintain.

Reply:
You know what would make me feel a lot better? If this could be packaged up in another package that cross-env could just consume...

Reply:
You know, I had the same thought. It's surely not the first time someone has wanted to parse bash style environment variables.

I'll take that on for the PEG version, but realistically it's likely to be a couple of weeks with other commitments and the shoulder I busted last week 😭.

Reply:
Oh no! Get well soon!

Reply:
FYI: https://github.com/franklin-ross/bash-env-parser

Reply:
This is pretty good to go right now. I've put a fair bit of effort into matching the behaviour of bash where it comes to the quotes, collapsing whitespace between words, etc. There are a bunch of tests which compare the library output against bash's output. I haven't pushed it up to npm yet, but you can install my latest commit from git and have a play if you'd like. It'd be good to get some feedback, if you have any.

Reply:
Thinking further, and looking at the `cross-env` code again, it feels like I should actually go the whole hog and parse the entire `VAR=$BLAH`. I was imagining you would be parsing that out and passing off some portion of the input to my package, but now that I think about it, it makes more sense to use a real parser for everything rather than the current regular expressions you've got. I basically find regex is fine until you need recursion and then it falls apart, which is the case if you want to add default value handling.

Am I on the right track? Biting off more than I should? I don't think it'd actually be a big deal for me to add, the hard part is done. I'm just concerned that integrating this will be more change than you're willing to make. Personally I think it's worth it for the default values, but it's your call.

Reply:
I like it!

Reply:
Gotta say I'm absolutely loving your new library. I'm not going to say I'm not a little sad about losing my state machine parser, but yours is much more elegant and covers more use cases.

I think the "VAR=value" parsing (i.e. to set the environment) seems like a good addition to your library as well though I feel like it's separate to most of the current functionality.

Have you gotten to the escape of the close curly brace in the default value handling? I read through most things and I can't seem to find it.

I think it makes sense to move a lot of the tests that are in this library into your new library as well.  Jest?  How do you want me to contribute? (you got a whole new viable public library on your hands!)

Reply:
That's great to hear, thanks! I've been wanting an excuse to build a parser for a while, so I'm happy to have found a good reason.

I agree that setting the value as part of the stringify function would be wrong, but my thought was to create a few different methods for walking the parse tree, one of which would be a more "interactive" style thing to set environment along the way. That's the next thing I'm gonna look at.

The brace matching should all be handled automatically by the grammar. There's something like this (I'm simplifying a bit) under the Variable production rule: `"${" VarName ":-" Fallback "}"`, and the Fallback can be 0 or more `Word / Quoted / Variable`, so they're mutually recursive and can nest indefinitely. If for some reason it fails to match then it falls back on just reading it as plain text, although bash refuses to parse, so I'm still on the fence about which way to go there.

Reply:
In terms of helping, I'd love that, but maybe in a little bit once I get things settled 😜. I'm still working pretty rough and ready at the moment, but hope to get it into a relatively stable state soon.

Reply:
Love it. I think this will drastically simplify cross-env (and any other library that attempts to do something with args like this).

Reply:
That's great, thanks a lot for the feedback! I sometimes find it difficult to tell whether my vision for something is as good as I think it is, or I'm just getting lost in the woods, so the validation is great. This feels pretty nice and clean though, so I'm happy to hear that.

I'll keep plugging away at it, and probably try and integrate it with a cross-env branch next week to see how that goes.

There's got to be somewhere better to talk about it than this PR though, but Github doesn't really have room for random chats. I might make a "cross-env integration" issue or something in the package that we can all continue talking on. Would that be better?

Reply:
We can continue here: https://github.com/franklin-ross/bash-env-parser/issues/7